stdlib/public/SDK/Foundation/JSONEncoder.swift

```diff
@@ -815,30 +823,45 @@ extension _JSONEncoder {
         }
     }
 
-    fileprivate func box<T : Encodable>(_ value: T) throws -> NSObject {
+    fileprivate func box(_ value: Encodable) throws -> NSObject {
         return try self.box_(value) ?? NSDictionary()
     }
 
+    fileprivate func box(_ dict: [String : Encodable]) throws {
+        let result = self.storage.pushKeyedContainer()
+        for (k, v) in dict {
+            self.codingPath.append(_JSONKey(stringValue: k, intValue: nil))
+            defer { self.codingPath.removeLast() }
+            result[k] = try box(v)
+        }
+        // Container will be popped by caller
+    }
+
     // This method is called "box_" instead of "box" to disambiguate it from the overloads. Because the return type here is different from all of the "box" overloads (and is more general), any "box" calls in here would call back into "box" recursively instead of calling the appropriate overload, which is not what we want.
-    fileprivate func box_<T : Encodable>(_ value: T) throws -> NSObject? {
-        if T.self == Date.self || T.self == NSDate.self {
+    fileprivate func box_(_ value: Encodable) throws -> NSObject? {
+        let type = type(of: value)
+        if type == Date.self || type == NSDate.self {
             // Respect Date encoding strategy
             return try self.box((value as! Date))
-        } else if T.self == Data.self || T.self == NSData.self {
+        } else if type == Data.self || type == NSData.self {
             // Respect Data encoding strategy
             return try self.box((value as! Data))
-        } else if T.self == URL.self || T.self == NSURL.self {
+        } else if type == URL.self || type == NSURL.self {
             // Encode URLs as single strings.
             return self.box((value as! URL).absoluteString)
-        } else if T.self == Decimal.self || T.self == NSDecimalNumber.self {
+        } else if type == Decimal.self || type == NSDecimalNumber.self {
             // JSONSerialization can natively handle NSDecimalNumber.
             return (value as! NSDecimalNumber)
         }
```

Comment:
We can probably move up the box call from below to up here. Performance-wise, I don't know how expensive it would be to check `if let stringKeyedDict = value as? [String : Encodable]` vs. adding another marker protocol and doing the protocol check, but this would be something good to test.

Reply:
I've added another marker protocol. I don't know how to go about testing the performance of the cast vs. a protocol check, but now the marker is there and the cast is only performed if the marker is present.